File: 054_spiral_matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> eles = new ArrayList<>();
        if (matrix.length == 0) return eles;
        int m = matrix.length, n = matrix[0].length;
        int row = 0, col = -1;
        while (true){
            // move right
            for (int i = 0; i < n; i++){
                eles.add(matrix[row][++col]);
            }
            if (--m == 0) break; // --m 代表 row 的數量扣一，扣到零為止代表已經走完

            // move down
            for (int i = 0; i < m; i++){
                eles.add(matrix[++row][col]);
            }
            if (--n == 0) break; // --n 代表 col 的數量扣一，扣到零為止代表已經走完

            // move left
            for (int i = 0; i < n; i++){
                eles.add(matrix[row][--col]);
            }
            if (--m == 0) break;

            // move up
            for (int i = 0; i < m; i++){
                eles.add(matrix[--row][col]);
            }
            if (--n == 0) break;
        }
        return eles;
    }
}

File: python/054_spiral_matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        eles = []
        if not matrix: return eles
        m, n = len(matrix), len(matrix[0])
        row, col = 0, -1

        while True:
            # move right
            for i in range(0, n):
                col += 1
                eles.append(matrix[row][col])
            m -= 1
            if m == 0: break

            # move down
            for i in range(0, m):
                row += 1
                eles.append(matrix[row][col])
            n -= 1
            if n == 0: break

            # move left
            for i in range(0, n):
                col -= 1
                eles.append(matrix[row][col])
            m -= 1
            if m == 0: break

            # move up
            for i in range(0, m):
                row -= 1
                eles.append(matrix[row][col])
            n -= 1
            if n == 0: break
        return eles

